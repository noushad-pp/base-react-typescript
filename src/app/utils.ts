import { AnyAction } from 'redux';
import { ApiRequestAction } from './request';

export const hasDefinedResponse = (status: keyof ApiRequestAction['responses'], action: ApiRequestAction): boolean => {
  return Boolean(action.responses[status]);
};

export const getResponse = (status: number, success: boolean, action: ApiRequestAction): AnyAction => {
  if (hasDefinedResponse(status, action)) {
    return action.responses[status];
  }

  if (success) {
    return action.responses.success;
  }

  return action.responses.error;
};
