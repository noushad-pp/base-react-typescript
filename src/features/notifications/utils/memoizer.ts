export default function memoizer<T, U>(): (key: T, val: U) => U {
  const map = new Map<T, U>();
  return (key: T, val: U): U => {
    if (!map.has(key)) {
      map.set(key, val);
    }

    return map.get(key)!;
  };
}
