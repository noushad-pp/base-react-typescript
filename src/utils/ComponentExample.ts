import { ReactElement } from 'react';

export class ComponentExample {
  public readonly name: string;
  public readonly code: ReactElement<any>;
  constructor(name: string, code: ReactElement<any>) {
    this.name = name;
    this.code = code;
  }
}
