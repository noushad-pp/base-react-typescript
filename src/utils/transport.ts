import { merge } from 'lodash';
import { AllowedMethods } from '../app/request';

export const defaultOptions: RequestInit = {
  credentials: 'include',
  headers: {
    'x-requested-with': 'XMLHttpRequest',
    'content-type': 'application/json',
    accept: 'application/json',
  },
};

export type ApiResponse = {
  response: unknown;
  success: boolean;
  status: number;
  statusText: string;
};

const doFetch = async (url: string, options: RequestInit) => fetch(url, merge(defaultOptions, options));

export const apiRequest = async <Response>(url: string, options: RequestInit): Promise<ApiResponse> => {
  const response = await doFetch(url, options);

  // TODO: handle non JSON response?
  const apiResponse: ApiResponse = {
    success: response.ok,
    status: response.status,
    statusText: response.statusText,
    response: await response.clone().json(),
  };

  return new Promise<ApiResponse>((resolve) => resolve(apiResponse));
};

export const request = async (method: AllowedMethods, url: string, body?: unknown, options: RequestInit = {}) =>
  apiRequest(url, { ...options, body: body !== undefined ? JSON.stringify(body) : undefined, method });
