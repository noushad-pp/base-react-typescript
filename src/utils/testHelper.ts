import fs from 'fs';
import path from 'path';
import React from 'react';
import { ComponentExample } from './ComponentExample';

const EXAMPLE_REGEXP = /\.example\.tsx$/;
const EXAMPLE_PATH = './src';

export function* findFiles(regexp: RegExp, currentPath: string = EXAMPLE_PATH): any {
  const realPath = path.resolve(currentPath);

  if (!fs.existsSync(realPath)) {
    return;
  }

  const files = fs.readdirSync(realPath);
  for (const name of files) {
    const fileName = path.join(realPath, name);
    const stat = fs.lstatSync(fileName);
    if (stat.isDirectory()) {
      yield* findFiles(regexp, fileName);
    }
    if (regexp.test(fileName)) {
      yield fileName;
    }
  }
}

export function* findExamples(currentPath: string = EXAMPLE_PATH): any {
  yield* findFiles(EXAMPLE_REGEXP, currentPath);
}

export function getAppRelativePath(filePath: string): string {
  return path.relative(process.cwd(), filePath);
}

type Components = {
  [s: string]: React.FunctionComponent | ComponentExample[];
};
type SupportedComponents = {
  [s: string]: React.FunctionComponent;
};

export const resolveComponents = (components: Components): SupportedComponents => {
  return Object.entries(components).reduce((carry: SupportedComponents, [name, value]): SupportedComponents => {
    if (value instanceof Function) {
      carry[name] = value;
    } else {
      value.forEach((example: unknown) => {
        if (!(example instanceof ComponentExample)) {
          throw new Error(`Only examples with instance of ComponentExample are allowed now.`);
        }
      });

      value.forEach((example: ComponentExample) => {
        if (carry[example.name]) {
          throw new Error(`Please use different names for your examples. Duplicated name: "${example.name}".`);
        }
        carry[example.name] = () => example.code;
      });
    }

    return carry;
  }, {});
};
