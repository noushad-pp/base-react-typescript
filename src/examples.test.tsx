import path from 'path';
import React from 'react';
import ReactDOM from 'react-dom';
import ReactDOMServer from 'react-dom/server';
import { findExamples, resolveComponents } from './utils/testHelper';

const examples = [...findExamples()];

examples.forEach((file) => {
  const fileName = path.basename(file);
  describe(`Examples ${fileName}`, () => {
    it('renders without errors', async () => {
      const components = await import(file);
      Object.entries(resolveComponents(components)).map(([name, Component]) => {
        const div = document.createElement('div');
        ReactDOM.render(<Component />, div);
        ReactDOM.unmountComponentAtNode(div);

        expect(ReactDOMServer.renderToString(<Component />)).toMatchSnapshot(name);
      });
    });
  });
});
