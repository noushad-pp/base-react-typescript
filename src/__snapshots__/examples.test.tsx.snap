// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Examples NotificationList.example.tsx renders without errors: Simple message 1`] = `"<div><div class=\\"successBar\\"><div class=\\"text\\">oh yes</div><i class=\\"close\\"></i></div><div class=\\"warningBar\\"><div class=\\"text\\">oh no</div><i class=\\"close\\"></i></div><div class=\\"errorBar\\"><div class=\\"text\\">oh shi</div><i class=\\"close\\"></i></div></div>"`;
